sdk/core/Azure.Core/src/ResourceIdentifier.cs

```diff
@@ -39,137 +64,178 @@ public sealed class ResourceIdentifier : IEquatable<ResourceIdentifier>, ICompar
+            if (resourceTypeName.Equals(SubscriptionsKey.AsSpan(), StringComparison.OrdinalIgnoreCase) && parent.ResourceType == ResourceType.Tenant)
+            {
+                specialType = SpecialType.Subscription;
+                return ResourceType.Subscription;
+            }
 
-            int validLength = string.IsNullOrEmpty(parts[parts.Length - 1]) ? parts.Length - 1 : parts.Length;
-            return AppendNext(Root, parts, 1, validLength);
+            specialType = resourceTypeName.Equals(LocationsKey.AsSpan(), StringComparison.OrdinalIgnoreCase) ? SpecialType.Location : null;
+            return parent.ResourceType.AppendChild(resourceTypeName.ToString());
         }
 
-        private static ResourceIdentifier AppendNext(ResourceIdentifier parent, string[] parts, int startIndex, int length)
+#pragma warning disable CA2208 // Instantiate argument exceptions correctly
+        private static ResourceIdentifierParts GetNextParts(ResourceIdentifier parent, ref ReadOnlySpan<char> remaining, ref ReadOnlySpan<char> nextWord)
         {
-            int partsCount = length - startIndex;
-
-            if (partsCount == 0)
-                return parent;
-
-            string firstElement = parts[startIndex];
+            ReadOnlySpan<char> firstWord = nextWord;
+            ReadOnlySpan<char> secondWord = PopNextWord(ref remaining);
 
-            if (partsCount == 1)
+            if (secondWord.IsEmpty)
             {
                 //subscriptions and resourceGroups aren't valid ids without their name
-                if (firstElement.Equals(SubscriptionsKey, StringComparison.OrdinalIgnoreCase) || firstElement.Equals(ResourceGroupsLowerKey, StringComparison.OrdinalIgnoreCase))
-                    throw new ArgumentOutOfRangeException(nameof(parts), $"The ResourceIdentifier is missing the key for {firstElement}.");
+                if (firstWord.Equals(SubscriptionsKey.AsSpan(), StringComparison.OrdinalIgnoreCase) || firstWord.Equals(ResourceGroupKey.AsSpan(), StringComparison.OrdinalIgnoreCase))
+                    throw new ArgumentOutOfRangeException("resourceId", $"The ResourceIdentifier is missing the key for {firstWord.ToString()}.");
 
                 //resourceGroup must contain either child or provider resource type
-                if (parent.ResourceType == ResourceGroupResourceType)
-                    throw new ArgumentOutOfRangeException(nameof(parts), $"Expected {ProvidersKey} path segment after {ResourceGroupsLowerKey}.");
+                if (parent.ResourceType == ResourceType.ResourceGroup)
+                    throw new ArgumentOutOfRangeException("resourceId", $"Expected {ProvidersKey} path segment after {ResourceGroupKey}.");
 
-                return new ResourceIdentifier(parent, firstElement, string.Empty);
+                nextWord = secondWord;
+                SpecialType? specialType = firstWord.Equals(LocationsKey.AsSpan(), StringComparison.OrdinalIgnoreCase) ? SpecialType.Location : null;
```

Comment:
This would box in the case of `null` being set. It can be avoided by using a 'None' value.